src/server/network/network.ts

```diff
@@ -0,0 +1,152 @@
+                    address:
+                      ipv4.length > 0 ? ipv4[0].address : ipv6[0].address,
+                    port: team.port,
+                  },
+                  (
+                    team: TeamData,
+                    payload: Buffer,
+                    rinfo: dgram.RemoteInfo,
+                  ) => {
+                    this.onMessage(team, payload, rinfo);
+                  },
+                ),
+              );
+            }
+          }
+        },
+      );
+    });
+  }
+
+  static of(teams: Array<TeamData>) {
+    return new UDPServers(teams);
+  }
+
+  // Add a new client to our list of clients
+  on(client: string, emit_cb: ClientListener) {
+    // If the client is already in our list, remove it
+    if (this.clients.has(client)) {
+      this.clients.delete(client);
+    }
+
+    // Add the client
+    this.clients.set(client, emit_cb);
+
+    // Return the off callback
+    return () => this.clients.delete(client);
+  }
+
+  // One of the UDP servers received a new message, send it on to all clients
+  private onMessage(team: TeamData, packet: Buffer, rinfo: dgram.RemoteInfo) {
+    for (let emit_cb of this.clients.values()) {
+      emit_cb('udp_packet', {
+        payload: packet,
+        team: {
+          name: team.name,
+          address: team.address,
+          port: team.port,
+        },
+        rinfo: {
+          family: rinfo.family,
+          address: rinfo.address,
+          port: rinfo.port,
+        },
+      });
+    }
+  }
+}
+
+class UDPServer {
+  private socket: dgram.Socket;
+
+  constructor(private team: TeamData, private msg_cb: MessageCallback) {
+    this.socket = dgram.createSocket({
+      type: ip6addr.parse(team.address).kind() === 'ipv6' ? 'udp6' : 'udp4',
+      reuseAddr: true,
+    });
+
+    this.socket.on('listening', () => {
+      // Add multicast membership if we are using a multicast address
+      if (
+        ip6addr.createCIDR('224.0.0.0/4').contains(team.address) ||
+        ip6addr.createCIDR('ff00::/8').contains(team.address)
+      ) {
+        this.socket.addMembership(team.address);
+      }
+      const address = this.socket.address();
+      console.log(
+        `${team.name} UDP Server: Listening on ${address.address}:${address.port}`,
+      );
+    });
+
+    this.socket.on('error', (error: Error) => {
+      console.log(`${team.name} UDP Server: Error:\n${error.stack}`);
+      this.socket.close();
+    });
+
+    this.socket.on('message', (packet: Buffer, rinfo: dgram.RemoteInfo) => {
+      console.log(
+        `${team.name} UDP Server: Received a message from ${rinfo.family}:${rinfo.address}:${rinfo.port}`,
+      );
+      this.msg_cb(this.team, packet, rinfo);
+    });
+
+    this.socket.on('close', () => {
+      console.log('${team.name} UDP Server: Closed');
+    });
+
+    this.socket.bind({ port: team.port, address: team.address });
+  }
+
+  static of(team: TeamData, msg_cb: MessageCallback) {
```

Comment:
Do we really need the `of()` pattern here? We use it in NUsight to provide a simple construction interface to classes that need many dependencies to construct. Same with the Network class. Perhaps @BrendanAnnable has some thoughts.